projects/igniteui-angular/src/lib/grids/grid-base.directive.ts

```diff
@@ -6677,12 +6677,11 @@ export abstract class IgxGridBaseDirective extends DisplayDensityBase implements
      */
     protected autogenerateColumns() {
         const data = this.gridAPI.get_data();
-        const factory = this.resolver.resolveComponentFactory(IgxColumnComponent);
         const fields = this.generateDataFields(data);
         const columns = [];
 
         fields.forEach((field) => {
-            const ref = factory.create(this.viewRef.injector);
+            const ref = this.viewRef.createComponent(IgxColumnComponent, { injector: this.injector, ngModuleRef: this.moduleRef });
             ref.instance.field = field;
             ref.instance.dataType = this.resolveDataTypes(data[0][field]);
             ref.changeDetectorRef.detectChanges();
```

Comment:
@rkaraivanov Can you tell me if the `detectChanges()` call is really necessary here. If it's needed, I suppose calling it once outside of the loop would be better.

Reply:
AFAIK it is needed since column components are OnPush and we are changing the initial input values (I recall that Angular did not call ngOnChanges with dynamic components and OnPush, but that could be fixed).
As for the `detectChanges()` call, it is on the component instance itself, not on the host view, so I'm not entirely sure if we can batch it in a single call. Could be tested though..